desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+    """Roots of cubic polynomial.
+
+    Parameters
+    ----------
+    coef : ndarray
+        First axis should store coefficients of a polynomial. For a polynomial
+        given by c₁ x³ + c₂ x² + c₃ x + c₄, ``coef[i]`` should store cᵢ.
+        It is assumed that c₁ is nonzero.
+    constant : ndarray, shape(constant.size, )
+        Specify to instead find solutions to c₁ x³ + c₂ x² + c₃ x + c₄ = ``constant``.
+    a_min, a_max : ndarray
+        Minimum and maximum value to clip roots between.
+        If None, clipping is not performed on the corresponding edge.
+        Otherwise, complex roots are clipped to ``a_min`` which defaults to -infinity.
+        Both arrays are broadcast against arrays of shape ``coef.shape[1:]``.
+    sort : bool
+        Whether to sort the roots.
+
+    Returns
+    -------
+    roots : ndarray, shape(constant.size, coef.shape, 3)
+        If constant has one element, the first axis will be squeezed out.
+        The roots of the cubic polynomial.
+
+    """
+    # https://en.wikipedia.org/wiki/Cubic_equation#General_cubic_formula
+    # The common libraries use root-finding which isn't compatible with JAX.
+    clip = not (a_min is None and a_max is None)
+    if a_min is None:
+        a_min = -jnp.inf
+    if a_max is None:
+        a_max = jnp.inf
+
+    a, b, c, d = coef
+    d = jnp.squeeze((d[jnp.newaxis].T - constant).T)
+    t_0 = b**2 - 3 * a * c
+    t_1 = 2 * b**3 - 9 * a * b * c + 27 * a**2 * d
+    C = ((t_1 + complex_sqrt(t_1**2 - 4 * t_0**3)) / 2) ** (1 / 3)
+    C_is_zero = jnp.isclose(C, 0)
+
+    def compute_roots(xi_k):
+        t_3 = jnp.where(C_is_zero, 0, t_0 / (xi_k * C))
+        r = -(b + xi_k * C + t_3) / (3 * a)
+        return r
+
+    def clip_roots(r):
+        r = jnp.where(jnp.isreal(r), jnp.clip(jnp.real(r), a_min, a_max), a_min)
+        return r
+
+    xi_1 = (-1 + (-3) ** 0.5) / 2
+    xi_2 = xi_1**2
+    xi_3 = 1
+    roots = tuple(map(compute_roots, (xi_1, xi_2, xi_3)))
+    if clip:
+        roots = tuple(map(clip_roots, roots))
+    roots = jnp.stack(roots, axis=-1)
+    if sort:
+        roots = jnp.sort(roots, axis=-1)
+    return roots
+
+
+def _get_bounce_points(pitch, zeta, poly_B, poly_B_z):
+    """Get the bounce points given |B| and 1 / λ.
+
+    Parameters
+    ----------
+    pitch : ndarray
+        λ values representing the constant function 1 / λ.
+    zeta : ndarray
+        Field line-following ζ coordinates of spline knots.
+    poly_B : ndarray
+        Polynomial coefficients of the cubic spline of |B|.
+    poly_B_z : ndarray
+        Polynomial coefficients of the cubic spline of ∂|B|/∂_ζ.
+
+    Returns
+    -------
+    intersect, is_bp, bp1, bp2 : ndarray, ndarray, ndarray, ndarray
+        The polynomials' intersection points with 1 / λ is given by ``intersect``.
+        In order to be JIT compilable, the returned array must have a shape that
+        accommodates the case where each cubic polynomial intersects 1 / λ thrice.
```

Comment:
if we use monotonic splines it should guarantee only 1 intersection per interval, though might be less/more accurate in some cases

Reply:
In that case there would be 0 or 1 intersection per interval. The number of bounce integrals we want to do depends on the supplied `pitch` values. The algorithm needs to be agnostic of the `pitch` values to be differentiable with jit, so there's no way to avoid a potentially sparse array.

Reply:
Yeah it wouldn't help with the sparsity but could reduce the array sizes by 3x. 

Though if you're specifying the derivatives of the spline exactly then it won't generally be monotonic so might be moot.